lib/src/main/java/space/taran/arklib/domain/meta/RootMetadataStorage.kt

```diff
@@ -0,0 +1,167 @@
+import space.taran.arklib.utils.LogTags.METADATA
+import java.lang.IllegalStateException
+import java.nio.file.Path
+import kotlin.io.path.*
+
+class RootMetadataStorage(
+    private val index: RootIndex,
+    private val appScope: CoroutineScope
+) : MetadataStorage {
+    val root = index.path
+
+    private val _inProgress = MutableStateFlow(false)
+    private val _updates = MutableSharedFlow<MetadataUpdate>()
+
+    private val metadataDir = root.arkFolder().arkMetadata()
+
+    private fun metadataPath(id: ResourceId): Path =
+        metadataDir.resolve(id.toString())
+
+    init {
+        Log.d(METADATA, "Initializing metadata storage for root $root")
+        metadataDir.createDirectories()
+        initUpdatedResourcesListener()
+        initKnownResources()
+    }
+
+    override val inProgress = _inProgress.asStateFlow()
+
+    override val updates = _updates.asSharedFlow()
+
+    fun locate(id: ResourceId): Result<Metadata> {
+        val metadataPath = metadataPath(id)
+        if (!metadataPath.exists()) {
+            return Result.failure(
+                IllegalStateException("Metadata for $id doesn't exist")
+            )
+        }
+
+        return readMetadata(metadataPath)
+    }
+
+    override fun locate(path: Path, id: ResourceId): Result<Metadata> =
+        locate(id)
+
+    override suspend fun forget(id: ResourceId) {
+        _updates.emit(MetadataUpdate.Deleted(id))
+        metadataPath(id).deleteIfExists()
+    }
+
+    private fun generate(resources: Collection<NewResource>) {
+        val amount = resources.size
+        Log.i(METADATA, "Checking metadata for $amount known resources in $root")
+
+        appScope.launch(Dispatchers.IO) {
+            _inProgress.emit(true)
+
+            val jobs = resources.map { added ->
+                launch { generate(added.path, added.resource) }
+            }
+
+            jobs.joinAll()
+            _inProgress.emit(false)
+        }
+    }
+
+    private suspend fun generate(path: Path, resource: Resource) {
+        require(!path.isDirectory()) { "Folders are not allowed here" }
+        Log.d(METADATA,
+            "Generating metadata for resource ${resource.id} by path $path")
+
+        val locator = metadataPath(resource.id)
+
+        MetadataGenerator
+            .generate(path, resource)
+            .onSuccess { metadata ->
+                val json = when (metadata) {
+                    is Metadata.Archive -> Json.encodeToString(metadata)
+                    is Metadata.Document -> Json.encodeToString(metadata)
+                    is Metadata.Image -> Json.encodeToString(metadata)
+                    is Metadata.Link -> Json.encodeToString(metadata)
+                    is Metadata.PlainText -> Json.encodeToString(metadata)
+                    is Metadata.Video -> Json.encodeToString(metadata)
+                }
+
+                locator.writeText(json)
+                _updates.emit(MetadataUpdate.Added(resource.id, path, metadata))
```

Comment:
I think it's better to send updates in batches, like in an `Index`
Since `PreviewStorage` processes changes from `MetadataStorage`, it will often update `inProgress`, which means that ui will "lag"

Reply:
I agree, we should change it. Right now, it was difficult.

Reply:
For this moment, `PreviewProcessor` (see #56) batches individual `Added` updates and updates `busy` flow only once for the batch.

Reply:
By the way, for asynchronous processing individual updates are better, because it's possible to do different stages at the same moment (metadata is being extracted for some resources while for some other resources previews are being generated). Probably, progress indicators should be separate for each resource (e.g. spinning circles on icons).